src/orange/g4org/SolidConverter.cc

```diff
@@ -656,8 +656,24 @@ auto SolidConverter::trap(arg_type solid_base) -> result_type
 auto SolidConverter::trd(arg_type solid_base) -> result_type
 {
     auto const& solid = dynamic_cast<G4Trd const&>(solid_base);
-    CELER_DISCARD(solid);
-    CELER_NOT_IMPLEMENTED("trd");
+
+    auto hz = scale_(solid.GetZHalfLength());
+    auto hy1 = scale_(solid.GetYHalfLength1());
+    auto hy2 = scale_(solid.GetYHalfLength2());
+    auto hx1 = scale_(solid.GetXHalfLength1());
+    auto hx2 = scale_(solid.GetXHalfLength2());
+
+    std::vector<GenTrap::Real2> lower(4), upper(4);
+    lower[0] = GenTrap::Real2{-hx1, -hy1};
+    lower[1] = GenTrap::Real2{+hx1, -hy1};
+    lower[2] = GenTrap::Real2{+hx1, +hy1};
+    lower[3] = GenTrap::Real2{-hx1, +hy1};
+    upper[0] = GenTrap::Real2{-hx2, -hy2};
+    upper[1] = GenTrap::Real2{+hx2, -hy2};
+    upper[2] = GenTrap::Real2{+hx2, +hy2};
+    upper[3] = GenTrap::Real2{-hx2, +hy2};
+
+    return make_shape<GenTrap>(solid, hz, lower, upper);
```

Comment:
I'd prefer (for the sake of reusability and testability) for the work of constructing these points be moved into a `GenTrap` static helper function that takes the half-height, the lower half-widths, and the upper half-widths and returns a `GenTrap`:

```suggestion
    return make_shape<GenTrap>(solid, GenTrap::from_rectangles(hz,{hx1, hy1}, {hx2, hy2}));
```

and add a test to `ConvexRegions.test.cc` with this constructor

Reply:
GenTrap in Geant4 potentially has twisted sides!  It is an extension of the standard trap, which contains only flat sides.

Reply:
Yep! I guess my point is that this `GenTrap` convex region is now general enough to encompass several object types... "generic trapezoid" with flat faces is the first one; the code here removes the need for an "optimized" `Trd` shape. And honestly the way it's implemented, it could be any sort of generic polygon with flat top and bottom...

Reply:
Regarding your suggestion to write more constructors with different algorithms, do you intend to map Geant4 constructors one-to-one? 

For instance, your proposal above (in green, line 666) is appropriate for the G4Trd, which has [this constructor](https://gitlab.cern.ch/geant4/geant4/-/blob/master/source/geometry/solids/CSG/include/G4Trd.hh#L66).  But your helper function is only useful for the TRD.  So are you proposing to implement one helper function per G4 trap-like constructor?

Then we'll need another helper function for the [full G4Trap constructor](https://gitlab.cern.ch/geant4/geant4/-/blob/master/source/geometry/solids/CSG/include/G4Trap.hh#L114), with 3 angles _in addition_ to 7 half-lengths (hz, hy1, hy2, hx1, hx2, hx3, hx4) -- is that the idea?

Reply:
Aha, sorry I missed that there are three different trapezoids. Yes I think it will improve testing and reuse to have a distinct constructor for each of the cases.

Reply:
(But to clarify: we don't need a constructor for every Geant4 constructor: was just want one for each of the trapezoid "types".)

Reply:
Oh wait, G4Trap is the "genTrap". Maybe to avoid further confusion we should meet briefly on zoom...

Reply:
I will defer the full G4Trap constructor for later, after the twisted planes are implemented.